### ISSUE TITLE
StaticLinkageCheck false positive for test libraries?

### ISSUE THREAD

**Issue by suztomo** [GoogleCloudPlatform/cloud-opensource-java#291]
StaticLinkageChecker shows errors on grpc-testing.

```
grpc-testing-1.17.1.jar (10 errors):
  Linked from: io.grpc:grpc-testing:1.17.1
  ClassSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=com.google.common.truth.Truth}
  ClassSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=com.google.common.truth.StringSubject}
  ClassSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=com.google.common.truth.LongSubject}
  MethodSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=com.google.common.truth.Truth, methodName=assertThat, descriptor=(Ljava/lang/String;)Lcom/google/common/truth/StringSubject;}
  MethodSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=com.google.common.truth.StringSubject, methodName=isEqualTo, descriptor=(Ljava/lang/Object;)V}
  MethodSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=com.google.common.truth.StringSubject, methodName=matches, descriptor=(Ljava/lang/String;)V}
  MethodSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=com.google.common.truth.Truth, methodName=assertThat, descriptor=(Ljava/lang/Long;)Lcom/google/common/truth/LongSubject;}
  MethodSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=com.google.common.truth.LongSubject, methodName=isGreaterThan, descriptor=(Ljava/lang/Comparable;)V}
  MethodSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=com.google.common.truth.LongSubject, methodName=isEqualTo, descriptor=(Ljava/lang/Object;)V}
  MethodSymbolReference{sourceClassName=io.grpc.internal.testing.AbstractTransportTest, targetClassName=org.junit.Assert, methodName=assertNotEquals, descriptor=(Ljava/lang/Object;Ljava/lang/Object;)V}

```

They're mostly related to Truth testing library. https://search.maven.org/artifact/io.grpc/grpc-testing/1.17.1/jar has dependency to the library with latest version with scope:test :

```
    <dependency>
      <groupId>com.google.truth</groupId>
      <artifactId>truth</artifactId>
      <version>0.42</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>io.grpc</groupId>
      <artifactId>grpc-testing-proto</artifactId>
      <version>1.17.1</version>
      <scope>test</scope>
    </dependency>
  </dependencies>
```

The problem can be reproduced with following test:

```
  @Test
  public void testArtifactsToPaths_testLibrary() throws RepositoryException {

    Artifact grpcArtifact = new DefaultArtifact("io.grpc:grpc-testing:1.17.1");
    ListMultimap<Path, DependencyPath> multimap =
        StaticLinkageChecker.artifactsToPaths(ImmutableList.of(grpcArtifact));

    Set<Path> paths = multimap.keySet();

    Truth.assertThat(paths)
        .comparingElementsUsing(PATH_FILE_NAMES)
        .contains("truth-0.42.jar");
  }
```

```
<[/usr/local/google/home/suztomo/.m2/repository/io/grpc/grpc-testing/1.17.1/grpc-testing-1.17.1.jar, /usr/local/google/home/suztomo/.m2/repository/io/grpc/grpc-core/1.17.1/grpc-core-1.17.1.jar, /usr/local/google/home/suztomo/.m2/repository/io/grpc/grpc-stub/1.17.1/grpc-stub-1.17.1.jar, /usr/local/google/home/suztomo/.m2/repository/junit/junit/4.12/junit-4.12.jar, /usr/local/google/home/suztomo/.m2/repository/org/mockito/mockito-core/1.9.5/mockito-core-1.9.5.jar, /usr/local/google/home/suztomo/.m2/repository/io/grpc/grpc-context/1.17.1/grpc-context-1.17.1.jar, /usr/local/google/home/suztomo/.m2/repository/com/google/code/gson/gson/2.7/gson-2.7.jar, /usr/local/google/home/suztomo/.m2/repository/com/google/errorprone/error_prone_annotations/2.2.0/error_prone_annotations-2.2.0.jar, /usr/local/google/home/suztomo/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar, /usr/local/google/home/suztomo/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.17/animal-sniffer-annotations-1.17.jar, /usr/local/google/home/suztomo/.m2/repository/com/google/guava/guava/26.0-android/guava-26.0-android.jar, /usr/local/google/home/suztomo/.m2/repository/io/opencensus/opencensus-api/0.17.0/opencensus-api-0.17.0.jar, /usr/local/google/home/suztomo/.m2/repository/io/opencensus/opencensus-contrib-grpc-metrics/0.17.0/opencensus-contrib-grpc-metrics-0.17.0.jar, /usr/local/google/home/suztomo/.m2/repository/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar, /usr/local/google/home/suztomo/.m2/repository/org/hamcrest/hamcrest-core/1.1/hamcrest-core-1.1.jar, /usr/local/google/home/suztomo/.m2/repository/org/objenesis/objenesis/1.0/objenesis-1.0.jar, /usr/local/google/home/suztomo/.m2/repository/org/checkerframework/checker-compat-qual/2.5.2/checker-compat-qual-2.5.2.jar, /usr/local/google/home/suztomo/.m2/repository/com/google/errorprone/error_prone_annotations/2.1.3/error_prone_annotations-2.1.3.jar, /usr/local/google/home/suztomo/.m2/repository/com/google/j2objc/j2objc-annotations/1.1/j2objc-annotations-1.1.jar, /usr/local/google/home/suztomo/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.14/animal-sniffer-annotations-1.14.jar, /usr/local/google/home/suztomo/.m2/repository/io/grpc/grpc-context/1.14.0/grpc-context-1.14.0.jar, /usr/local/google/home/suztomo/.m2/repository/com/google/guava/guava/20.0/guava-20.0.jar, /usr/local/google/home/suztomo/.m2/repository/jmock/jmock/1.1.0/jmock-1.1.0.jar, /usr/local/google/home/suztomo/.m2/repository/junit/junit/4.0/junit-4.0.jar, /usr/local/google/home/suztomo/.m2/repository/org/easymock/easymock/2.2/easymock-2.2.jar, /usr/local/google/home/suztomo/.m2/repository/com/google/code/findbugs/jsr305/1.3.9/jsr305-1.3.9.jar, /usr/local/google/home/suztomo/.m2/repository/com/google/errorprone/error_prone_annotations/2.0.12/error_prone_annotations-2.0.12.jar, /usr/local/google/home/suztomo/.m2/repository/junit/junit/3.8.1/junit-3.8.1.jar]>
```

# Options

- Include dependencies with test scope into the class path
- grpc-testing to remove `<scope>test</scope>` for Truth library
- Add Truth (latest version 0.42) to Cloud OSS BOM

**Comment by elharo** [comment]
Or should io.grpc:grpc-testing:1.17.1 have scope test?

**Comment by suztomo** [comment]
Grpc team fixed this in 1.20.
